tsdb/engine/tsm1/engine.go

```diff
@@ -972,45 +1022,34 @@ func (e *Engine) compactTSMFull() {
 						files []string
 						err   error
 					)
-					if optimize {
-						files, err = e.Compactor.CompactFast(group)
-						if err == errCompactionsDisabled || err == errCompactionInProgress {
-							e.logger.Printf("aborted %s compaction group (%d). %v",
-								logDesc, groupNum, err)
-
-							if err == errCompactionInProgress {
-								time.Sleep(time.Second)
-							}
-							return
-						} else if err != nil {
-							e.logger.Printf("error compacting TSM files: %v", err)
-							atomic.AddInt64(&e.stats.TSMOptimizeCompactionErrors, 1)
 
-							time.Sleep(time.Second)
-							return
-						}
-					} else {
-						files, err = e.Compactor.CompactFull(group)
-						if err == errCompactionsDisabled || err == errCompactionInProgress {
-							e.logger.Printf("aborted %s compaction group (%d). %v",
-								logDesc, groupNum, err)
-
-							if err == errCompactionInProgress {
-								time.Sleep(time.Second)
-							}
-							return
-						} else if err != nil {
-							e.logger.Printf("error compacting TSM files: %v", err)
-							atomic.AddInt64(&e.stats.TSMFullCompactionErrors, 1)
+					// Count the compaction as active only while the compaction is actually running.
+					func() {
```

Comment:
I agree that the other one involved a lot of indentation, but I think the new one adds too much indirection now too. What I meant by nil interfaces is you could make an interface for each of the methods that need to be called and then have structs that implement the interfaces. The interfaces don't have to have any functionality. You can do something like this:

```
type Compaction interface {
    ... methods needed ...
}

type compactOptimize struct{}

func (*compactOptimize) Compact() { ... }

func main() {
    compacter := (*compactOptimize)(nil)
}
```

That way you can write the implementations on the top level scope with no indentation and to choose one, you just select one of the interfaces so it groups all of the methods together and they are clearly labeled for what they do. I don't know if this is necessary, but I think it might reduce the amount of indentation.